File: Scripts/scan_va_election_json.py
# Script to scan va_county_aggregated_results.json and extract insights for research findings
# Usage: Run with your .venv Python (see below)
# Example: .venv/Scripts/python.exe scan_va_election_json.py

import json
import os
from collections import defaultdict

DATA_PATH = os.path.join('Data', 'va_county_aggregated_results.json')

# Helper to calculate margin
def get_margin(dem, rep):
    total = dem + rep
    if total == 0:
        return 0
    return ((dem - rep) / total) * 100

def scan_trends(election_data):
    results = {}
    by_year = election_data.get('results_by_year', {})
    years = sorted(by_year.keys())
    # Use correct contest keys as in JSON
    key_races = ['President', 'Governor', 'US Senate']
    if len(years) < 2:
        return results
    last_year = years[-1]
    prev_year = years[-2]
    results['largest_margins'] = {}
    results['biggest_swings'] = {}
    results['statewide_margins'] = {}
    results['flips'] = {}
    for race in key_races:
        last = by_year.get(last_year, {}).get(race, {})
        prev = by_year.get(prev_year, {}).get(race, {})
        max_dem = {'county': '', 'margin': float('-inf')}
        max_rep = {'county': '', 'margin': float('inf')}
        biggest_swing = {'county': '', 'swing': 0}
        for county, l in last.items():
            p = prev.get(county, {})
            margin_last = get_margin(l.get('dem_votes', 0), l.get('rep_votes', 0))
            margin_prev = get_margin(p.get('dem_votes', 0), p.get('rep_votes', 0))
            if margin_last > max_dem['margin']:
                max_dem = {'county': county, 'margin': margin_last}
            if margin_last < max_rep['margin']:
                max_rep = {'county': county, 'margin': margin_last}
            swing = margin_last - margin_prev
            if abs(swing) > abs(biggest_swing['swing']):
                biggest_swing = {'county': county, 'swing': swing}
        results['largest_margins'][race] = {'year': last_year, 'max_dem': max_dem, 'max_rep': max_rep}
        results['biggest_swings'][race] = {'year': last_year, 'biggest_swing': biggest_swing}
        # Statewide margin trend
        margins = []
        for year in years:
            race_data = by_year.get(year, {}).get(race, {})
            dem = sum(row.get('dem_votes', 0) for row in race_data.values())
            rep = sum(row.get('rep_votes', 0) for row in race_data.values())
            margins.append({'year': year, 'margin': get_margin(dem, rep)})
        results['statewide_margins'][race] = margins
        # Flips
        flips = []
        for county, l in last.items():
            p = prev.get(county, {})
            margin_last = get_margin(l.get('dem_votes', 0), l.get('rep_votes', 0))
            margin_prev = get_margin(p.get('dem_votes', 0), p.get('rep_votes', 0))
            if (margin_last > 0 and margin_prev < 0) or (margin_last < 0 and margin_prev > 0):
                flips.append({'county': county, 'from': 'Dem' if margin_prev > 0 else 'Rep', 'to': 'Dem' if margin_last > 0 else 'Rep'})
        results['flips'][race] = {'year': last_year, 'flips': flips}
    return results

def main():
    with open(DATA_PATH, 'r', encoding='utf-8') as f:
        data = json.load(f)
    insights = scan_trends(data)
    print(json.dumps(insights, indent=2))

if __name__ == '__main__':
    main()

File: Scripts/scan_va_election_json.js
// Script to scan va_county_aggregated_results.json and extract insights for research findings
// Run with: node scan_va_election_json.js

const fs = require('fs');
const path = require('path');

const DATA_PATH = path.join(__dirname, 'Data', 'va_county_aggregated_results.json');

function loadElectionData() {
  const raw = fs.readFileSync(DATA_PATH, 'utf8');
  return JSON.parse(raw);
}

function getMargin(dem, rep) {
  const total = dem + rep;
  if (total === 0) return 0;
  return ((dem - rep) / total) * 100;
}

function scanTrends(electionData) {
  const results = [];
  const byYear = electionData.results_by_year || {};
  const years = Object.keys(byYear).sort();
  const keyRaces = ['president', 'governor', 'us_senate'];

  // 1. Find counties with largest Dem and Rep margins in recent years
  const lastYear = years[years.length - 1];
  const prevYear = years[years.length - 2];
  if (!lastYear || !prevYear) return results;

  keyRaces.forEach(race => {
    const last = byYear[lastYear][race] || {};
    const prev = byYear[prevYear][race] || {};
    let maxDem = { county: '', margin: -Infinity };
    let maxRep = { county: '', margin: Infinity };
    let biggestSwing = { county: '', swing: 0 };
    Object.keys(last).forEach(county => {
      const l = last[county];
      const p = prev[county] || {};
      const marginLast = getMargin(l.dem_votes || 0, l.rep_votes || 0);
      const marginPrev = getMargin(p.dem_votes || 0, p.rep_votes || 0);
      if (marginLast > maxDem.margin) maxDem = { county, margin: marginLast };
      if (marginLast < maxRep.margin) maxRep = { county, margin: marginLast };
      const swing = marginLast - marginPrev;
      if (Math.abs(swing) > Math.abs(biggestSwing.swing)) biggestSwing = { county, swing };
    });
    results.push({
      race,
      year: lastYear,
      maxDem,
      maxRep,
      biggestSwing
    });
  });

  // 2. Statewide margin trend for each race
  keyRaces.forEach(race => {
    const margins = years.map(year => {
      const raceData = byYear[year][race] || {};
      let dem = 0, rep = 0;
      Object.values(raceData).forEach(row => {
        dem += row.dem_votes || 0;
        rep += row.rep_votes || 0;
      });
      return { year, margin: getMargin(dem, rep) };
    });
    results.push({ race, statewideMargins: margins });
  });

  // 3. Counties that flipped party between last two cycles
  keyRaces.forEach(race => {
    const last = byYear[lastYear][race] || {};
    const prev = byYear[prevYear][race] || {};
    const flips = [];
    Object.keys(last).forEach(county => {
      const l = last[county];
      const p = prev[county] || {};
      const marginLast = getMargin(l.dem_votes || 0, l.rep_votes || 0);
      const marginPrev = getMargin(p.dem_votes || 0, p.rep_votes || 0);
      if ((marginLast > 0 && marginPrev < 0) || (marginLast < 0 && marginPrev > 0)) {
        flips.push({ county, from: marginPrev > 0 ? 'Dem' : 'Rep', to: marginLast > 0 ? 'Dem' : 'Rep' });
      }
    });
    if (flips.length) results.push({ race, year: lastYear, flips });
  });

  return results;
}

function main() {
  const data = loadElectionData();
  const insights = scanTrends(data);
  console.log(JSON.stringify(insights, null, 2));
}

main();
